server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -464,6 +477,82 @@ func (r *statsRecorder) Stop() {
 	close(r.onStatsRecorded)
 }
 
+func persistArtifact(ctx context.Context, env environment.Env, uri *url.URL, path string) error {
+	w, err := env.GetBlobstore().Writer(ctx, path)
+	if err != nil {
+		return status.WrapErrorf(
+			err,
+			"Failed to open writer to blobstore for path %s to persist cache artifact at %s",
+			path,
+			uri.String(),
+		)
+	}
+	if err := bytestream.StreamBytestreamFile(ctx, env, uri, w); err != nil {
+		w.Close()
+		return status.WrapErrorf(
+			err,
+			"Failed to stream to blobstore for path %s to persist cache artifact at %s",
+			path,
+			uri.String(),
+		)
+	}
+	if err := w.Commit(); err != nil {
+		w.Close()
+		return status.WrapErrorf(
+			err,
+			"Failed to commit to blobstore for path %s to persist cache artifact at %s",
+			path,
+			uri.String(),
+		)
+	}
+	if err := w.Close(); err != nil {
+		return status.WrapErrorf(
+			err,
+			"Failed to close blobstore writer for path %s to persist cache artifact at %s",
+			path,
+			uri.String(),
+		)
+	}
+	return nil
+}
+
+func persistTestDataForInvocation(ctx context.Context, env environment.Env, inv *inpb.Invocation, path string) {
+	for _, e := range inv.GetEvent() {
+		p, ok := e.BuildEvent.Payload.(*build_event_stream.BuildEvent_TestResult)
+		if !ok {
+			continue
+		}
+		for _, f := range p.TestResult.TestActionOutput {
+			resultURI, err := url.Parse(f.GetUri())
+			if err != nil {
+				log.Errorf("Error parsing URI for test result: %s", f.GetUri())
+				continue
+			}
+			if resultURI.Scheme != "bytestream" {
+				continue
+			}
+			if _, ok := testArtifactsToPersist[f.Name]; !ok {
+				continue
+			}
+			target := strings.TrimPrefix(e.BuildEvent.Id.GetTestResult().Label, "//")
+			if i := strings.LastIndex(target, ":"); i != -1 {
+				target = target[:i] + "/" + target[i:]
+			}
```

Comment:
I think doing this path manipulation is error prone, and there are all kinds of edge cases that you'd have to account for i.e. test with multiple runs, tests with multiple shards, retries, etc.

You wouldn't have to do a fallback for every request if you used cache paths - you could just take a boolean flag on whether or not to fallback to blobstore, and we could set that in the UI on places where there might be a fallback (i.e. test logs and timing profiles for now).

Reply:
Not sure if this is what you meant @siggisim but instead of using the path at all, we could use the SHA256 of the test artifact as the blobstore path. e.g. if the file name is bytestream://remote.buildbuddy.io/blob/abc123/456 then we could store it in blobstore under `{invocation_id}/artifacts/sha256/abc123` -- would that work?

Reply:
switched to cache-style paths.